File: leetcode/easy/226_invert_binary_tree.js
// Invert a binary tree.

// Recursively; Time Complexity: O(n) and Space Complexity: O(n) since the stack frames.

var invertTree = function(root) {
  if (!root) return null;
  
  let dfs = (node) => {
    if (!node) return null;
    let right = dfs(node.right);
    let left = dfs(node.left);
    
    node.left = right;
    node.right = left;
    
    return node;
  }
  
  dfs(root);
  return root;
};

// Itteratively

var invertTree = function(root) {
  const queue = [root]
  
  while (queue.length) {
    let node = queue.shift();
    if (node) {
      [node.left, node.right] = [node.right, node.left];
      queue.push(node.left, node.right);
    }
  }
  
  return root;
};

// Time Complexity: O(n) and space is O(n);

// Example:

// Input:

//      4
//    /   \
//   2     7
//  / \   / \
// 1   3 6   9
// Output:

//      4
//    /   \
//   7     2
//  / \   / \
// 9   6 3   1

File: python/trees/226_invertbinarytree.py
# Invert a binary tree.

# Example:

# Input:

#      4
#    /   \
#   2     7
#  / \   / \
# 1   3 6   9
# Output:

#      4
#    /   \
#   7     2
#  / \   / \
# 9   6 3   1
# Trivia:
# This problem was inspired by this original tweet by Max Howell:

# Google: 90% of our engineers use the software you wrote (Homebrew), but you can’t invert a binary tree on a whiteboard so f*** off.

# Recursive

# class Solution:
#     def invertTree(self, root: TreeNode) -> TreeNode:
#         if root is None:
#           return None
#         right = self.invertTree(root.right)
#         left = self.invertTree(root.left)
#         root.left = right
#         root.right = left
#         return root

#Itterative
# class Solution:
#     def invertTree(self, root: TreeNode) -> TreeNode:
#       if root is None:
#         return None
      
#       queue = [root]
#       while len(queue):
#         curr_node = queue.pop()
#         temp = curr_node.left
#         curr_node.left = curr_node.right
#         curr_node.right = temp
        
#         if curr_node.left != None: queue.append(curr_node.left)
#         if curr_node.right != None: queue.append(curr_node.right)
      
#       return root